sycl/plugins/unified_runtime/pi2ur.hpp

```diff
@@ -1918,21 +1920,33 @@ inline pi_result piextKernelSetArgMemObj(pi_kernel Kernel, pi_uint32 ArgIndex,
   if (ArgValue)
     UrMemory = reinterpret_cast<ur_mem_handle_t>(*ArgValue);
 
-  ur_kernel_arg_mem_obj_properties_t Properties{};
-
   // We don't yet know the device where this kernel will next be run on.
   // Thus we can't know the actual memory allocation that needs to be used.
   // Remember the memory object being used as an argument for this kernel
   // to process it later when the device is known (at the kernel enqueue).
   //
-  // TODO: for now we have to conservatively assume the access as read-write.
-  //       Improve that by passing SYCL buffer accessor type into
-  //       piextKernelSetArgMemObj.
-  //
-
   ur_kernel_handle_t UrKernel = reinterpret_cast<ur_kernel_handle_t>(Kernel);
-  HANDLE_ERRORS(
-      urKernelSetArgMemObj(UrKernel, ArgIndex, &Properties, UrMemory));
+  if (ArgProperties) {
+    assert(ArgProperties->type == PI_KERNEL_ARG_MEM_OBJ_ACCESS);
+    ur_kernel_arg_mem_obj_properties_t UrMemProperties{};
+    UrMemProperties.stype = UR_STRUCTURE_TYPE_KERNEL_ARG_MEM_OBJ_PROPERTIES;
+    switch (ArgProperties->mem_access) {
+    case PI_ACCESS_READ_ONLY:
+      UrMemProperties.memoryAccess = UR_MEM_FLAG_READ_ONLY;
+      break;
+    case PI_ACCESS_WRITE_ONLY:
+      UrMemProperties.memoryAccess = UR_MEM_FLAG_WRITE_ONLY;
+      break;
+    case PI_ACCESS_READ_WRITE:
+      UrMemProperties.memoryAccess = UR_MEM_FLAG_READ_WRITE;
+      break;
+    }
+    HANDLE_ERRORS(
+        urKernelSetArgMemObj(UrKernel, ArgIndex, &UrMemProperties, UrMemory));

```

Comment:
can't we define the `pi_mem_obj_property` in the same way as `ur_kernel_arg_mem_obj_properties_t` to not need pi2ur conversion here and just pass through?

Reply:
I think so, I am worrying that it will require reinterpret casts that is hardly debugged in case of UR APIs changes.  I will update it.

Reply:
@smaslov-intel done here https://github.com/intel/llvm/pull/9752/commits/4584ea1e0a331e8f615e9b672478544c578cdc8b